qmail-showctl.8
.TH qmail-showctl 8
.SH NAME
qmail-showctl \- analyze the qmail configuration files
.SH SYNOPSIS
.B qmail-showctl
.SH DESCRIPTION
.B qmail-showctl
explains the current
.B qmail
configuration.
.SH "SEE ALSO"
qmail-control(8)


CHANGES
20190714 code: removed idedit, install-big. var-qmail packages are no more.
20190708 code: lookup uid/gid at runtime.
20190711 portability problem: explicitly initialize strerr_sys and
         define BIND_8_COMPAT to work around Mac OS X.
20190709 doc: rename INSTALL, SENDMAIL to INSTALL.md, SENDMAIL.md to
         support building on case-insensitive filesystems.
20190715 code: detect and prefer utmpx where available.
20190708 code: use DESTDIR environment variable as root directory in install.
20071130 version: netqmail 1.06
20071130 legal: qmail-1.03 is now in the public domain
20051103 doc: dot-qmail.9 updated for changed (19980613) conf-patrn default, tnx ADM
20040121 version: netqmail 1.05
20040121 code: qmail-smtpd is protected from exceedingly long (eg 2GB)
         header lines
20040121 code: qmail_lspawn, qmail-newmrh, qmail-newu, and qmail-rspawn
         are protected from misbehaving on hosts where the size of an
         integer is not the same as the size of a character pointer
         (eg 64 bit hosts with 32 bit ints)
20031027 version: netqmail 1.04.
20031027 doc: INSTALL points to http://lifewithqmail.org/lwq.html
20031027 doc: qmail.7 identifies installation as netqmail and points to
         http://qmail.org/
20031027 doc: qmail-queue.8 adds explanation of $QMAILQUEUE
20031027 doc: qmail-log.5 adds reference to errors from $QMAILQUEUE script
20031027 doc: FAQ also points to http://cr.yp.to/qmail/faq.html and
         http://qmail.org/
20031027 code: qmail-smtpd identifies itself as netqmail
20031027 code: if $QMAILQUEUE is set, it's invoked instead of qmail-queue
20031024 code: changed errno from int to #include.
20031024 code: fixed .qmail parsing bug.
20031024 code: recognize 0.0.0.0 as a local address.
20031024 code: sendmail's -f flag now overrides environment variables.
19980615 version: qmail 1.03.
19980614 doc: eliminated BIN.setup in favor of a web page.
19980614 code: added other auto* to qmail-showctl output.
19980614 doc: added pointer to immhf.html in qmail-header.5.
19980614 doc: added note to TEST.receive about SMTP command format.
19980614 doc: added FAQ 5.6 on qmail-qmqpd.
19980614 code: removed unused variables in idedit.c.
19980613 code: changed conf-patrn to 002.
19980613 doc: moved SENDMAIL lower in INSTALL.
19980612 code: added install-big.
19980612 code: added BIN.Makefile.
19980612 doc: added BIN.README, BIN.setup.
19980612 code: switched to new install.
19980611 code: added idedit.
19980611 doc: added FAQ 1.3 on $QMAILMFTFILE.
19980611 doc: used bouncesaying in FAQ 5.5.
19980611 code: added except.
19980611 code: added bouncesaying.
19980611 code: allowed unbracketed IP addresses in dns_ipplus() and
         dns_mxip().
19980611 code: allowed spaces after colon in non-bracketed addresses in
         qmail-smtpd.
19980611 doc: cleaned up UPGRADE.
19980528 bug: qmail-smtpd skips first character in rcpthosts() call.
         tnx NND. impact: qmail-smtpd crashes on empty address; and it
	 allows relaying to ""@any.host. fix: use addr.s.
19980515 doc: expanded flock discussion in INSTALL.mbox.
19980515 doc: eliminated flock warning from INSTALL.maildir.
19980515 doc: split REMOVE.binmail out of INSTALL.
19980515 doc: split REMOVE.sendmail out of INSTALL.
19980515 doc: split TEST.deliver and TEST.receive out of INSTALL and
         UPGRADE.
19980515 doc: integrated INSTALL.boot into INSTALL.
19980515 code: cleaned up final output in qmail-qmqpd.c.
19980514 doc: updated procmail notes in INSTALL.mbox. tnx JRM.
19980514 doc: changed FAQ 4.4 to point to INSTALL.mbox for procmail.
         tnx JRM.
19980514 code: separated HELO and EHLO; single-line response for HELO.
         tnx to various people.
19980430 version: qmail 1.02.
19980430 doc: updated SECURITY.
19980430 doc: fixed FAQ 4.9. tnx KB.
19980430 code: changed quote2() to avoid quoting <>.
19980429 code: changed quote_need() to quote empty local parts. tnx HHO.
19980428 doc: added status notes to INSTALL and UPGRADE.
19980428 code: skip setting environment in sendmail.c if PROTO is set.
19980428 code: eliminated recipientmap.
19980428 code: added virtual users to qmail-send.c. tnx RN.
19980428 code: eliminated domain from rewrite() in qmail-send.c.
19980428 code: added binm1, binm1+df, binm2, binm2+df, binm3, binm3+df.
19980428 doc: eliminated most Mailbox references from INSTALL, UPGRADE.
19980428 code: added config-fast.
19980428 code: renamed qmail-config as config.
19980428 code: supported QMAILMFTFILE in qmail-inject.c.
19980428 code: recognized Mail-Followup-To in hfield.c.
19980428 code: replaced rwrecip() with rwappend() in qmail-inject.c.
19980428 code: cleaned up doheaderfile() in qmail-inject.c.
19980426 code: eliminated -type test from qmail-qstat to speed it up.
         tnx FT.
19980421 doc: eliminated remove-rcpthosts comments from FAQ.
19980421 doc: updated FAQ 4.3 to point to Russ Allbery's FAQ. 
19980421 doc: took account of /var/qmail/boot in INSTALL, UPGRADE, and
         INSTALL.vsm.
19980421 code: added /var/qmail/boot, with home, home+df, proc, proc+df.
19980421 doc: skipped make and make man in INSTALL.
19980420 doc: cleaned up mbox description in SENDMAIL.
19980420 code: changed QMQP port to official port 628.
19980402 doc: updated qmsmac references to fastforward.
19980402 doc: replaced qmail-upgrade man page with doc/SENDMAIL.
19980402 code: added qmqpservers output to qmail-showctl.
19980402 code: added qmail-qmqpd.
19980402 code: added qmail-qmqpc.
19980304 code: eliminated del_saywhynoexit in qmail-send.c.
19980304 code: eliminated concurrencynodel in qmail-send.c.
19980222 code: added status() to qmail-send.c.
19980222 code: added concurrencyused to qmail-send.c.
19980128 doc: added note to qmail-getpw.9 about ETXTBSY.
19980127 code: eliminated err_seenmail() in qmail-smtpd.c. tnx PO.
19980126 doc: used $DEFAULT in FAQ where possible.
19980126 code: added DEFAULT in qmail-local.
19980126 code: added -/ to qmail-pw2u.

qmail-showctl.c
#include <sys/types.h>
#include <sys/stat.h>
#include "substdio.h"
#include "subfd.h"
#include "exit.h"
#include "fmt.h"
#include "str.h"
#include "control.h"
#include "constmap.h"
#include "stralloc.h"
#include "direntry.h"
#include "uidgid.h"
#include "auto_uids.h"
#include "auto_users.h"
#include "auto_qmail.h"
#include "auto_break.h"
#include "auto_patrn.h"
#include "auto_spawn.h"
#include "auto_split.h"

stralloc me = {0};
int meok;

stralloc line = {0};
char num[FMT_ULONG];

int auto_uida;
int auto_uidd;
int auto_uidl;
int auto_uido;
int auto_uidp;
int auto_uidq;
int auto_uidr;
int auto_uids;

int auto_gidn;
int auto_gidq;

void safeput(buf,len)
char *buf;
unsigned int len;
{
  char ch;

  while (len > 0) {
    ch = *buf;
    if ((ch < 32) || (ch > 126)) ch = '?';
    substdio_put(subfdout,&ch,1);
    ++buf;
    --len;
  }
}

void do_int(fn,def,pre,post)
char *fn;
char *def;
char *pre;
char *post;
{
  int i;
  substdio_puts(subfdout,"\n");
  substdio_puts(subfdout,fn);
  substdio_puts(subfdout,": ");
  switch(control_readint(&i,fn)) {
    case 0:
      substdio_puts(subfdout,"(Default.) ");
      substdio_puts(subfdout,pre);
      substdio_puts(subfdout,def);
      substdio_puts(subfdout,post);
      substdio_puts(subfdout,".\n");
      break;
    case 1:
      if (i < 0) i = 0;
      substdio_puts(subfdout,pre);
      substdio_put(subfdout,num,fmt_uint(num,i));
      substdio_puts(subfdout,post);
      substdio_puts(subfdout,".\n");
      break;
    default:
      substdio_puts(subfdout,"Oops! Trouble reading this file.\n");
      break;
  }
}

void do_str(fn,flagme,def,pre)
char *fn;
int flagme;
char *def;
char *pre;
{
  substdio_puts(subfdout,"\n");
  substdio_puts(subfdout,fn);
  substdio_puts(subfdout,": ");
  switch(control_readline(&line,fn)) {
    case 0:
      substdio_puts(subfdout,"(Default.) ");
      if (!stralloc_copys(&line,def)) {
	substdio_puts(subfdout,"Oops! Out of memory.\n");
	break;
      }
      if (flagme && meok)
	if (!stralloc_copy(&line,&me)) {
	  substdio_puts(subfdout,"Oops! Out of memory.\n");
	  break;
	}
    case 1:
      substdio_puts(subfdout,pre);
      safeput(line.s,line.len);
      substdio_puts(subfdout,".\n");
      break;
    default:
      substdio_puts(subfdout,"Oops! Trouble reading this file.\n");
      break;
  }
}

int do_lst(fn,def,pre,post)
char *fn;
char *def;
char *pre;
char *post;
{
  int i;
  int j;

  substdio_puts(subfdout,"\n");
  substdio_puts(subfdout,fn);
  substdio_puts(subfdout,": ");
  switch(control_readfile(&line,fn,0)) {
    case 0:
      substdio_puts(subfdout,"(Default.) ");
      substdio_puts(subfdout,def);
      substdio_puts(subfdout,"\n");
      return 0;
    case 1:
      substdio_puts(subfdout,"\n");
      i = 0;
      for (j = 0;j < line.len;++j)
	if (!line.s[j]) {
          substdio_puts(subfdout,pre);
          safeput(line.s + i,j - i);
          substdio_puts(subfdout,post);
          substdio_puts(subfdout,"\n");
	  i = j + 1;
	}
      return 1;
    default:
      substdio_puts(subfdout,"Oops! Trouble reading this file.\n");
      return -1;
  }
}

void main()
{
  DIR *dir;
  direntry *d;
  struct stat stmrh;
  struct stat stmrhcdb;

  auto_uida = inituid(auto_usera);
  auto_uidd = inituid(auto_userd);
  auto_uidl = inituid(auto_userl);
  auto_uido = inituid(auto_usero);
  auto_uidp = inituid(auto_userp);
  auto_uidq = inituid(auto_userq);
  auto_uidr = inituid(auto_userr);
  auto_uids = inituid(auto_users);

  auto_gidn = initgid(auto_groupn);
  auto_gidq = initgid(auto_groupq);

  substdio_puts(subfdout,"qmail home directory: ");
  substdio_puts(subfdout,auto_qmail);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"qmail alias directory: ");
  substdio_puts(subfdout,auto_qmail_alias);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"qmail binaries directory: ");
  substdio_puts(subfdout,auto_qmail_bin);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"qmail boot directory: ");
  substdio_puts(subfdout,auto_qmail_boot);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"qmail documents directory: ");
  substdio_puts(subfdout,auto_qmail_doc);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"qmail control directory: ");
  substdio_puts(subfdout,auto_qmail_control);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"qmail manpage directory: ");
  substdio_puts(subfdout,auto_qmail_man);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"qmail queue directory: ");
  substdio_puts(subfdout,auto_qmail_queue);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"qmail users directory: ");
  substdio_puts(subfdout,auto_qmail_users);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"user-ext delimiter: ");
  substdio_puts(subfdout,auto_break);
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"paternalism (in decimal): ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_patrn));
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"silent concurrency limit: ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_spawn));
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"subdirectory split: ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_split));
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"user ids: ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uida));
  substdio_puts(subfdout,", ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidd));
  substdio_puts(subfdout,", ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidl));
  substdio_puts(subfdout,", ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uido));
  substdio_puts(subfdout,", ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidp));
  substdio_puts(subfdout,", ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidq));
  substdio_puts(subfdout,", ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidr));
  substdio_puts(subfdout,", ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uids));
  substdio_puts(subfdout,".\n");

  substdio_puts(subfdout,"group ids: ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_gidn));
  substdio_puts(subfdout,", ");
  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_gidq));
  substdio_puts(subfdout,".\n");

  if (chdir(auto_qmail_control) == -1) {
    substdio_puts(subfdout,"Oops! Unable to chdir to ");
    substdio_puts(subfdout,auto_qmail_control);
    substdio_puts(subfdout,".\n");
    substdio_flush(subfdout);
    _exit(111);
  }

  dir = opendir(".");
  if (!dir) {
    substdio_puts(subfdout,"Oops! Unable to open current directory.\n");
    substdio_flush(subfdout);
    _exit(111);
  }

  meok = control_readline(&me,"me");
  if (meok == -1) {
    substdio_puts(subfdout,"Oops! Trouble reading control/me.");
    substdio_flush(subfdout);
    _exit(111);
  }

  do_lst("badmailfrom","Any MAIL FROM is allowed.",""," not accepted in MAIL FROM.");
  do_str("bouncefrom",0,"MAILER-DAEMON","Bounce user name is ");
  do_str("bouncehost",1,"bouncehost","Bounce host name is ");
  do_int("concurrencylocal","10","Local concurrency is ","");
  do_int("concurrencyremote","20","Remote concurrency is ","");
  do_int("databytes","0","SMTP DATA limit is "," bytes");
  do_str("defaultdomain",1,"defaultdomain","Default domain name is ");
  do_str("defaulthost",1,"defaulthost","Default host name is ");
  do_str("doublebouncehost",1,"doublebouncehost","2B recipient host: ");
  do_str("doublebounceto",0,"postmaster","2B recipient user: ");
  do_str("envnoathost",1,"envnoathost","Presumed domain name is ");
  do_str("helohost",1,"helohost","SMTP client HELO host name is ");
  do_str("idhost",1,"idhost","Message-ID host name is ");
  do_str("localiphost",1,"localiphost","Local IP address becomes ");
  do_lst("locals","Messages for me are delivered locally.","Messages for "," are delivered locally.");
  do_str("me",0,"undefined! Uh-oh","My name is ");
  do_lst("percenthack","The percent hack is not allowed.","The percent hack is allowed for user%host@",".");
  do_str("plusdomain",1,"plusdomain","Plus domain name is ");
  do_lst("qmqpservers","No QMQP servers.","QMQP server: ",".");
  do_int("queuelifetime","604800","Message lifetime in the queue is "," seconds");

  if (do_lst("rcpthosts","SMTP clients may send messages to any recipient.","SMTP clients may send messages to recipients at ","."))
    do_lst("morercpthosts","No effect.","SMTP clients may send messages to recipients at ",".");
  else
    do_lst("morercpthosts","No rcpthosts; morercpthosts is irrelevant.","No rcpthosts; doesn't matter that morercpthosts has ",".");
  /* XXX: check morercpthosts.cdb contents */
  substdio_puts(subfdout,"\nmorercpthosts.cdb: ");
  if (stat("morercpthosts",&stmrh) == -1)
    if (stat("morercpthosts.cdb",&stmrhcdb) == -1)
      substdio_puts(subfdout,"(Default.) No effect.\n");
    else
      substdio_puts(subfdout,"Oops! morercpthosts.cdb exists but morercpthosts doesn't.\n");
  else
    if (stat("morercpthosts.cdb",&stmrhcdb) == -1)
      substdio_puts(subfdout,"Oops! morercpthosts exists but morercpthosts.cdb doesn't.\n");
    else
      if (stmrh.st_mtime > stmrhcdb.st_mtime)
        substdio_puts(subfdout,"Oops! morercpthosts.cdb is older than morercpthosts.\n");
      else
        substdio_puts(subfdout,"Modified recently enough; hopefully up to date.\n");

  do_str("smtpgreeting",1,"smtpgreeting","SMTP greeting: 220 ");
  do_lst("smtproutes","No artificial SMTP routes.","SMTP route: ","");
  do_int("timeoutconnect","60","SMTP client connection timeout is "," seconds");
  do_int("timeoutremote","1200","SMTP client data timeout is "," seconds");
  do_int("timeoutsmtpd","1200","SMTP server data timeout is "," seconds");
  do_lst("virtualdomains","No virtual domains.","Virtual domain: ","");

  while (d = readdir(dir)) {
    if (str_equal(d->d_name,".")) continue;
    if (str_equal(d->d_name,"..")) continue;
    if (str_equal(d->d_name,"bouncefrom")) continue;
    if (str_equal(d->d_name,"bouncehost")) continue;
    if (str_equal(d->d_name,"badmailfrom")) continue;
    if (str_equal(d->d_name,"bouncefrom")) continue;
    if (str_equal(d->d_name,"bouncehost")) continue;
    if (str_equal(d->d_name,"concurrencylocal")) continue;
    if (str_equal(d->d_name,"concurrencyremote")) continue;
    if (str_equal(d->d_name,"databytes")) continue;
    if (str_equal(d->d_name,"defaultdomain")) continue;
    if (str_equal(d->d_name,"defaulthost")) continue;
    if (str_equal(d->d_name,"doublebouncehost")) continue;
    if (str_equal(d->d_name,"doublebounceto")) continue;
    if (str_equal(d->d_name,"envnoathost")) continue;
    if (str_equal(d->d_name,"helohost")) continue;
    if (str_equal(d->d_name,"idhost")) continue;
    if (str_equal(d->d_name,"localiphost")) continue;
    if (str_equal(d->d_name,"locals")) continue;
    if (str_equal(d->d_name,"me")) continue;
    if (str_equal(d->d_name,"morercpthosts")) continue;
    if (str_equal(d->d_name,"morercpthosts.cdb")) continue;
    if (str_equal(d->d_name,"percenthack")) continue;
    if (str_equal(d->d_name,"plusdomain")) continue;
    if (str_equal(d->d_name,"qmqpservers")) continue;
    if (str_equal(d->d_name,"queuelifetime")) continue;
    if (str_equal(d->d_name,"rcpthosts")) continue;
    if (str_equal(d->d_name,"smtpgreeting")) continue;
    if (str_equal(d->d_name,"smtproutes")) continue;
    if (str_equal(d->d_name,"timeoutconnect")) continue;
    if (str_equal(d->d_name,"timeoutremote")) continue;
    if (str_equal(d->d_name,"timeoutsmtpd")) continue;
    if (str_equal(d->d_name,"virtualdomains")) continue;
    substdio_puts(subfdout,"\n");
    substdio_puts(subfdout,d->d_name);
    substdio_puts(subfdout,": I have no idea what this file does.\n");
  }

  substdio_flush(subfdout);
  _exit(0);
}
